docs/ronn/knoc.1.ronn
[comment]: # ( -*- Mode: markdown; -*- )
knoc(1) -- invoke Kno interpeter
=========================================

## SYNOPSIS

`knoc` [<config>|<arg>...]

## DESCRIPTION

This command launches the Kno console/terminal program, a
read-eval-print loop which allows you to evaluate expressions and see
their values.

The arguments to `knoc` are either filenames or *config* entries
of the form `**setting**=*name*`.  The files are loaded and the config
entries processed in order of appearance and config entries may
specify multiple values for a setting.

When normally started, `knoc` uses the **editline** library to
provide simple command line editing and history functions. This can be
suppressed by the environment variables `NO_EDITLINE` or
`INSIDE_EMACS`.

Editline maintains a per-session input history of its own which can be
used in history commands. Knoc itself maintains a *value history*
which allows previously displayed objects or sub-objects to be
referenced in new commands. See [History references].

Some config entries may generate useful side-effects defined by either
the Kno core or modules loaded by the core. The config system and
standard options are described [kno_config](here). Some of the most
useful built-in config settings are:

* `DOTLOAD=yes|no`
  disables the loading of the `~/.knoconfig` file on startup; this can
  also be specified by setting the environment variable `KNO_SKIP_DOTLOAD` to
  anything but the empty string;

* `ENVCONFIG=yes|no`
  disables configuration information from the shell environment;

* `CONFIG=*config files*;...`:
  loads configuration information from *config files* on startup; this is
  done after command line configs are processed;

* `LOADPATH=*path*`:
  specifies a search path for modules, as a colon
  separated list of patterns where the '%' character is replaced by
  the module name. For example `/opt/mods/%/module.scm:/opt/mods/%:`
  causes the system to look for a module `foo/x` in
  `/opt/mods/foo/x/module.scm` and `/opt/mods/foo/x.scm`. Paths can be
  relative to the current directory by starting with '.' and can also
  include ~home directory references.

* `LOGLEVEL=*level*`:
  specifies the default loglevel for notifications
  where level is either a positive integer or symbolic name (like
  `NOTICE` or `DEBUG` or `INFO`).

* `LOADMOD=*module name*;...`:
  loads the Kno modules *modnames* during startup, making them
  available in the default environment. This is done after the
  `CONFIG` specifications are all processed;

* `LOADFILE=*paths*;...`: loads the files *paths* during startup. This
  happens *after* the `LOADMOD` configs are processed.

* `INIT=*expr*`: evaluates *expr* in the default environment; this
  happens after the `LOADFILE` configs are processed.


## DIAGNOSTICS

Returns 0 when exiting normally, 1 otherwise. It also displays a short
preamble on starting and reports its time and memory usage on exit.

## FILES

On startup, `knoc` normally reads an `.knoconfig` file from the user's
home directory (if it exists). It can include `MODULES` or `LOADFILE`
declarations. The loading of the `.knoconfig` file can be suppressed
via the environment variable `KNO_SKIP_DOTLOAD` or the config variable
`DOTLOAD`.

## Value history

`knoc` has a simple history mechanism which assigns identifiers to
some displayed objects. By default, this is done for all non-trivial
return results (excluding short strings, small numbers, etc). Other
code may also add objects to the history, making those objects easily
accessible through the console.

The general form of a history reference is `#**num**` or
`#**num**.*selector*`. For example, a function might return five long
strings (elided here for clarity):

````console
{ ;; #27
  "long string 1..." ;; #27.0
  "long string 2..." ;; #27.1
  "long string 3..." ;; #27.2
  "long string 4..." ;; #27.3
  "long string 5..." ;; #27.4
  } ;; #27
````

where subsequent expressions can use history references, e.g.:

````console
#|kno>| (length #27.0)
#|=>| 2049
#|kno>| (choice-size #27)
#|=>| 5
````

When output elision is on, items don't need to be actually displayed
to be accessible with history references, for example, a large choice
might be displayed:

````console
{ ;; #28
  "long string 1..." ;; #28.0
  "long string 2..." ;; #28.1
  "long string 3..." ;; #28.2
  "long string 4..." ;; #28.3
  "long string 5..." ;; #28.4
  ;;; 251/256 more items
  } ;; #28
````

displays the first five of 256 choice items, but any of the items can be
referred to in history references, e.g.
````console
#|kno>| (length #27.128)
#|=>| 1024
````

When mapping objects (slotmaps, OIDs, etc) are displayed, they follow
a similar convention except the selector is a slotid/key rather than a
number. For example:
````console
;; #29
\#[name "longidstring...." ;; #29.name
   description "longerdescription string"] ;; #29.description
   ````

shows an object whose values can be referred to as `#29.name` or
`#29.description`.

## ENVIRONMENT

As with most Kno programs, config variables can be set through
environment variables prefixed with `KNO_`, so `KNO_LOGLEVEL` will set
the default loglevel. This can be disabled by the config setting
`ENVCONFIG=no` or by setting the environment variable
`KNO_DISABLE_ENVCONFIG` to anything but the empty string.

# AUTHOR

Ken Haase <kh@beingmeta.com>

## SEE ALSO

knox(1), kno_config(8)



debian/changelog
framerd (3.9.1-13) unstable; urgency=medium

  * Fix packaging descriptions
  * git patch version framerd-3.9.1-31-g2af7e18

 -- Ken Haase <kh@beingmeta.com>  Tue, 17 May 2016 11:11:23 -0400

framerd (3.9.1-12) unstable; urgency=medium

  * Split out framerd into framerd-core and framerd-builtin (which provides framerd-modules)
  * git patch version framerd-3.9.1-29-ge3a33fa

 -- Ken Haase <kh@beingmeta.com>  Tue, 17 May 2016 10:58:54 -0400

framerd (3.9.1-11) unstable; urgency=low

  * More adjustments to libmysqlclient dependencies
  * git patch version framerd-3.9.1-25-g880479a

 -- Ken Haase <kh@beingmeta.com>  Sat, 14 May 2016 10:12:25 +0000

framerd (3.9.1-10) unstable; urgency=low

  * Fix libmysqlclient dependency
  * Save alternate booksystem files in config directory
  * git patch version framerd-3.9.1-23-g4def5c8

 -- Ken Haase <kh@beingmeta.com>  Sat, 14 May 2016 09:17:44 +0000

framerd (3.9.1-9) unstable; urgency=low

  * Add libzip2 dependency to packaging
  * Fix FramerD systemd configuration
  * git patch version framerd-3.9.1-18

 -- Ken Haase <kh@beingmeta.com>  Fri, 13 May 2016 14:47:16 +0000

framerd (3.9.1-8) unstable; urgency=low

  * Add optional arg to empty-string? to treat non-breaking spaces as content
  * git patch version framerd-3.9.1-15-gff06aeb

 -- Ken Haase <kh@beingmeta.com>  Fri, 13 May 2016 03:51:54 +0000

framerd (3.9.1-7) unstable; urgency=low

  * Fix bug in optimize code
  * git patch version framerd-3.9.1-12-g99c9b27

 -- Ken Haase <kh@beingmeta.com>  Sat, 30 Apr 2016 22:51:01 +0000

framerd (3.9.1-6) unstable; urgency=medium

  * Repackage for alternate libmagickwand under xenial
  * git patch version framerd-3.9.1-9-gad2c5fc

 -- Ken Haase <kh@beingmeta.com>  Fri, 29 Apr 2016 13:16:58 -0400

framerd (3.9.1-5) unstable; urgency=medium

  * Improve error capture and handling including interrupts
  * Add fdstatus and fdlog commands
  * git patch version framerd-3.9

 -- Ken Haase <kh@beingmeta.com>  Thu, 31 Mar 2016 07:01:55 -0400

framerd (3.9.0-4) unstable; urgency=medium

  * Fix build flags to remove needless dependencies
  * git patch version framerd-3.9-15-ge4f2745

 -- Ken Haase <kh@beingmeta.com>  Tue, 29 Mar 2016 17:35:42 -0400

framerd (3.9.0-3) unstable; urgency=low

  * Fix opcode handling bugs
  * git patch version framerd-3.9-6-gc9661e4

 -- Ken Haase <kh@beingmeta.com>  Tue, 29 Mar 2016 11:14:47 -0400

framerd (3.9.0-2) unstable; urgency=low

  * Fix git tag and executable heralds
  * git patch version framerd-3.9

 -- Ken Haase <kh@beingmeta.com>  Mon, 28 Mar 2016 21:13:27 -0400

framerd (3.9.0-1) unstable; urgency=low

  * Fix systemd scripts
  * Added fdstatus script
  * Fix pool-load to set error context
  * Fixes to opcode bugs
  * Bumped minor release number

 -- Ken Haase <kh@beingmeta.com>  Mon, 28 Mar 2016 16:55:24 -0400

framerd (3.8.4-63) unstable; urgency=low

  * Fix libbson/libmongoc dependencies
  * git patch verison framerd-3.8.4-177-gdb5c379

 -- Ken Haase <kh@beingmeta.com>  Sun, 27 Mar 2016 00:31:54 +0000

framerd (3.8.4-62) unstable; urgency=low

  * Fix tidy5 dependencies
  * git patch version framerd-3.8.4-176-g33eb981

 -- Ken Haase <kh@beingmeta.com>  Sat, 26 Mar 2016 14:25:11 +0000

framerd (3.8.4-61) unstable; urgency=low

  * Add smarter immediate type checking
  * Fix some opcode/optimize bugs
  * git patch version framerd-3.8.4-173-gdc93e81

 -- Ken Haase <kh@beingmeta.com>  Wed, 23 Mar 2016 21:33:36 -0400

framerd (3.8.4-60) unstable; urgency=low

  * Fixes to xmlparsing
  * Fix some applytest error cases (whoops)
  * Check irritant pointer when calling fd_err
  * Improve backtrace CSS
  * git patch version framerd-3.8.4-168-g8a0394e

 -- Ken Haase <kh@beingmeta.com>  Wed, 23 Mar 2016 18:21:37 -0400

framerd (3.8.4-59) unstable; urgency=low

  * Add handling of extpool lockfn
  * make DOM cleanup take more initiative
  * git patch version framerd-3.8.4-156-g43cb8ea

 -- Ken Haase <kh@beingmeta.com>  Sat, 19 Mar 2016 19:51:14 +0000

framerd (3.8.4-58) unstable; urgency=low

  * Add set-procedure-args!
  * Add functions for augmenting errors in ONERROR handlers
  * Fixes to optimize module
  * Wrap primitives execution in call contours to convert setmped errors
    to return errors
  * git patch version framerd-3.8.4-153-gcd07c91

 -- Ken Haase <kh@beingmeta.com>  Mon, 14 Mar 2016 18:26:35 -0400

framerd (3.8.4-57) unstable; urgency=low

  * Pass priority on to logger from logif+
  * git patch version framerd-3.8.4-147-gfb67432

 -- Ken Haase <kh@beingmeta.com>  Sat, 05 Mar 2016 22:12:07 -0500

framerd (3.8.4-56) unstable; urgency=low

  * Fix bug in JWT changes
  * git patch version framerd-3.8.4-145-gebeeed4

 -- Ken Haase <kh@beingmeta.com>  Fri, 04 Mar 2016 10:43:32 -0500

framerd (3.8.4-55) unstable; urgency=low

  * Add RSA algorithm to JWT
  * Improve and fix parts of JWT module
  * Add third arg (urish) to ->base64 functions to use - and _
    for + and /
  * git patch version framerd-3.8.4-143-gdb7869b

 -- Ken Haase <kh@beingmeta.com>  Fri, 04 Mar 2016 08:57:17 -0500

framerd (3.8.4-54) unstable; urgency=low

  * Record boot handler and status in FramerD run dir
  * git patch version framerd-3.8.4-132-g22819cc (pre commit)

 -- Ken Haase <kh@beingmeta.com>  Mon, 29 Feb 2016 23:08:03 +0000

framerd (3.8.4-53) unstable; urgency=low

  * Fixes to daemon/servlet management scripts
  * git patch version framerd-3.8.4-128

 -- Ken Haase <kh@beingmeta.com>  Sun, 28 Feb 2016 20:09:55 +0000

framerd (3.8.4-52) unstable; urgency=low

  * Fix set-cookie to not automatically add dot to cookie domains
  * Fixes and refinements to jwt and jwt/auth
  * Make parsetime handle non-string arguments
  * git patch version framerd-3.8.4-126

 -- Ken Haase <kh@beingmeta.com>  Fri, 26 Feb 2016 22:58:38 +0000

framerd (3.8.4-51) unstable; urgency=low

  * Remove uridecode handling of '+' escapes
  * git patch version framer-3.8.4-121

 -- Ken Haase <kh@beingmeta.com>  Tue, 23 Feb 2016 18:51:41 +0000


etc/knoc.el
;;; -*- Mode: emacs-lisp; lexical-binding: t; -*-
;;; knoc.el --- emacs mode for the Knoc console

;; Copyright (C) 2001-2016  beingmeta, inc

;; Author: Ken Haase <kh@beingmeta.com>
;; Keywords: lisp kno framerd
;; Version: 3.9.5

;;; Commentary:

;; This package provides an emacs interaction mode for knoc, the KNO
;; REPL (read-eval-print loop). It also declares various indentation
;; rules for KNO Scheme functions.

;;; Code:

(require 'cmuscheme)

(let ((st scheme-mode-syntax-table))
  (modify-syntax-entry ?\[ "(]" scheme-mode-syntax-table)
  (modify-syntax-entry ?\] ")[" scheme-mode-syntax-table)
  (modify-syntax-entry ?{ "(}" scheme-mode-syntax-table)
  (modify-syntax-entry ?} "){" scheme-mode-syntax-table)
  (modify-syntax-entry ?( "()" scheme-mode-syntax-table)
  (modify-syntax-entry ?) ")(" scheme-mode-syntax-table))


;; We do this because we don't want console windows to have infinite undo
(make-variable-buffer-local 'undo-limit)
;; The name of the KNO scheme module for a particular buffer
(make-variable-buffer-local 'knoc-module)
;; The initial code to send to the buffer
(make-variable-buffer-local 'knoc-startup)
;; The knoc command line
(make-variable-buffer-local 'knoc-cmdline)

(defvar *kno-keywords*
  '("\\<do-choices-mt\\>" "\\<do-vector-mt\\>" "\\<for-choices-mt\\>"
    "\\<do-choices\\>" "\\<for-choices\\>" "\\<filter-choices\\>"
    "\\<doseq\\>" "\\<dolist\\>" "\\<dotimes\\>" "\\<forseq>\\"
    "\\<lambda\\>" "\\<ambda\\>" "\\<slambda\\>" "\\<macro\\>"
    "\\<try\\>" "\\<tryif\\>" "\\<with-log-context\\>"
    "\\<while\\>" "\\<until\\>" "\\<onerror>\\" "\\<set+!>\\"
    "\\<find-frames\\>" "\\<pick\\>" "\\<reject\\>"
    "\\<div\\>" "\\<p\\>" "\\<p*\\>" "\\<form\\>"
    "\\<try-choices>\\" "\\<tryseq>\\" "\\<extdb/proc>\\"
    "\\<cond>\\" "\\<if>\\" "\\<and>\\" "\\<or>\\"
    "\\<xmlout>\\" "\\<printout>\\" "\\<fileout>\\"
    "\\<stringout>\\" "\\choice\\" "\\when\\"
    "\\unless\\" "\\begin\\" "\\let\\" "\\let*\\"
    "\\while\\" "\\until\\" "\\not\\"
    "\\union\\" "\\intersection\\" "\\difference\\"
    "\\pick\\" "\\reject\\"
    "\\<error>\\" "\\<irritant>\\"))

(defvar calculate-lisp-indent-last-sexp)

(defun scheme-indent-function (indent-point state)
  (let ((normal-indent (current-column)))
    (goto-char (1+ (elt state 1)))
    (parse-partial-sexp (point) calculate-lisp-indent-last-sexp 0 t)
    (if (and (elt state 2)
	     (or (not (looking-at "\\sw\\|\\s_"))
		 (save-excursion
		   (goto-char (elt state 1))
		   (or (looking-at "{") (looking-at "\\[")))
		 (save-excursion
		   (goto-char (- (elt state 1) 1))
		   (or (looking-at "#(")
		       (looking-at "#\\[")
		       (looking-at "#{"))))
             ;; (not (looking-at "\\sw\\|\\s_"))
	     )
        ;; car of form doesn't seem to be a symbol
        (progn
          (if (not (> (save-excursion (forward-line 1) (point))
		      calculate-lisp-indent-last-sexp))
              (progn (goto-char calculate-lisp-indent-last-sexp)
                     (beginning-of-line)
                     (parse-partial-sexp (point)
					 calculate-lisp-indent-last-sexp 0 t)))
          ;; Indent under the list or under the first sexp on the same
          ;; line as calculate-lisp-indent-last-sexp.  Note that first
          ;; thing on that line has to be complete sexp since we are
          ;; inside the innermost containing sexp.
          (backward-prefix-chars)
          (current-column))
      (let ((function (buffer-substring (point)
					(progn (forward-sexp 1) (point))))
	    method)
	(setq method (or (get (intern-soft function) 'scheme-indent-function)
			 (get (intern-soft function) 'scheme-indent-hook)))
	(cond ((or (eq method 'defun)
		   (and (null method)
			(> (length function) 3)
			(string-match "\\`def" function)))
	       (lisp-indent-defform state indent-point))
	      ((integerp method)
	       (lisp-indent-specform method state
				     indent-point normal-indent))
	      (method
		(funcall method state indent-point normal-indent)))))))

(defun block-indenter (state indent-point normal)
  (save-excursion
    (goto-char (elt state 1))
    (+ (current-column) 2)))

;;; KNO stuff
(put 'when 'scheme-indent-function 1)
(put 'unless 'scheme-indent-function 1)
(put 'tryif 'scheme-indent-function 1)
(put 'withenv 'scheme-indent-function 1)
(put 'withenv/safe 'scheme-indent-function 1)
(put 'with-log-context 'scheme-indent-function 1)
(put 'error 'scheme-indent-function 2)
(put 'error+ 'scheme-indent-function 2)
(put 'error 'scheme-indent-function 3)
(put 'error+ 'scheme-indent-function 3)
(put 'irritant 'scheme-indent-function 2)

(put 'opts+ 'scheme-indent-function 2)
(put 'opt+ 'scheme-indent-function 2)

(put 'ambda 'scheme-indent-function 1)
(put 'sambda 'scheme-indent-function 1)
(put 'slambda 'scheme-indent-function 1)

(put 'set! 'scheme-indent-function 1)
(put 'set+! 'scheme-indent-function 1)
(put 'store! 'scheme-indent-function 2)
(put 'add! 'scheme-indent-function 2)
(put 'drop! 'scheme-indent-function 2)
(put 'assert! 'scheme-indent-function 2)
(put 'retract! 'scheme-indent-function 2)
(put 'test 'scheme-indent-function 2)
(put '%test 'scheme-indent-function 2)

(put 'irritant 'scheme-indent-function 2)
(put 'begin 'scheme-indent-function 'block-indenter)

(put 'thread/call 'scheme-indent-function 2)
(put 'thread/call+ 'scheme-indent-function 2)

(put 'dolist 'scheme-indent-function 1)
(put 'do-pool 'scheme-indent-function 1)
(put 'dotimes 'scheme-indent-function 1)
(put 'doseq 'scheme-indent-function 1)
(put 'dopool 'scheme-indent-function 1)
(put 'do-choices 'scheme-indent-function 1)
(put 'do-subsets 'scheme-indent-function 1)
(put 'for-choices 'scheme-indent-function 1)
(put 'forseq 'scheme-indent-function 1)
(put 'filter-choices 'scheme-indent-function 1)
(put 'try-choices 'scheme-indent-function 1)
(put 'do-choices-mt 'scheme-indent-function 1)
(put 'do-vector-mt 'scheme-indent-function 1)
(put 'for-choices-mt 'scheme-indent-function 1)
(put 'tryseq 'scheme-indent-function 1)
(put 'while 'scheme-indent-function 1)
(put 'until 'scheme-indent-function 1)
(put 'with-lock 'scheme-indent-function 1)
(put 'onerror 'scheme-indent-function 2)
(put 'prog1 'scheme-indent-function 1)
(put 'config-def! 'scheme-indent-function 1)

(put 'hashtable-increment! 'scheme-indent-function 2)
(put 'hashtable-increment-existing! 'scheme-indent-function 2)
(put 'table-increment! 'scheme-indent-function 2)
(put 'table-increment-existing! 'scheme-indent-function 2)
(put 'hashtable-multiply! 'scheme-indent-function 2)
(put 'hashtable-multiply-existing! 'scheme-indent-function 2)
(put 'table-multiply! 'scheme-indent-function 2)
(put 'table-multiply-existing! 'scheme-indent-function 2)

(put 'unwind-protect 'scheme-indent-function 1)
(put 'on-errors 'scheme-indent-function 1)

(put 'printout-to 'scheme-indent-function 1)
(put 'hmac-sha1 'scheme-indent-function 1)
(put 'printout 'scheme-indent-function 'block-indenter)
(put 'message 'scheme-indent-function 'block-indenter)
(put 'lineout 'scheme-indent-function 'block-indenter)
(put 'stringout 'scheme-indent-function 'block-indenter)
(put 'glom 'scheme-indent-function 'block-indenter)
(put 'fileout 'scheme-indent-function 1)
(put 'writeout 'scheme-indent-function 1)
(put 'writeout/type 'scheme-indent-function 2)
(put 'gp/writeout 'scheme-indent-function 2)
(put 'gp/writeout! 'scheme-indent-function 2)
(put 'gp/write! 'scheme-indent-function 3)
(put 'gp/save! 'scheme-indent-function 2)
(put 'gp/copy! 'scheme-indent-function 2)
(put 's3/op 'scheme-indent-function 3)
(put 's3/write! 'scheme-indent-function 2)
(put 's3/copy! 'scheme-indent-function 2)
(put 'ses/call 'scheme-indent-function 1)
(put 'logif 'scheme-indent-function 1)
(put 'logger 'scheme-indent-function 1)
(put 'logpanic 'scheme-indent-function 1)
(put 'logalert 'scheme-indent-function 1)
(put 'logcrit 'scheme-indent-function 1)
(put 'logerr 'scheme-indent-function 1)
(put 'logerror 'scheme-indent-function 1)
(put 'logwarn 'scheme-indent-function 1)
(put 'lognotice 'scheme-indent-function 1)
(put 'loginfo 'scheme-indent-function 1)
(put 'logdebug 'scheme-indent-function 1)
(put 'logdetail 'scheme-indent-function 1)
(put 'logdeluge 'scheme-indent-function 1)
(put 'logswamp 'scheme-indent-function 1)

(put 'collection/get 'scheme-indent-function 2)
(put 'collection/find 'scheme-indent-function 2)
(put 'collection/count 'scheme-indent-function 2)
(put 'collection/update! 'scheme-indent-function 2)
(put 'collection/insert! 'scheme-indent-function 1)
(put 'collection/upsert! 'scheme-indent-function 1)
(put 'collection/remove! 'scheme-indent-function 2)
(put 'collection/modify 'scheme-indent-function 2)
(put 'collection/modify! 'scheme-indent-function 2)

(put 'mongo/cursor 'scheme-indent-function 2)
(put 'mongo/getcollection 'scheme-indent-function 2)
(put 'mongo/results 'scheme-indent-function 2)
(put 'mongo/do 'scheme-indent-function 2)

(put 'mongovec 'scheme-indent-function 'block-indenter)

(put 'sig/make 'scheme-indent-function 1)
(put 'sig/check 'scheme-indent-function 2)
(put 'sig/check/ 'scheme-indent-function 3)

(put 'set-cookie! 'scheme-indent-function 1)
(put 'req/set! 'scheme-indent-function 1)
(put 'req/add! 'scheme-indent-function 1)

(put 'with-output 'scheme-indent-function 1)
(put 'with-output-to-string 'scheme-indent-function 'block-indenter)
(put 'string-subst* 'scheme-indent-function 1)

(put 'make-pool 'scheme-indent-function 1)
(put 'make-index 'scheme-indent-function 1)
(put 'open-pool 'scheme-indent-function 1)
(put 'open-index 'scheme-indent-function 1)
(put 'engine/run 'scheme-indent-function 2)

;;; XML/HTML generation
(put 'xmlout 'scheme-indent-function 'block-indenter)
(put 'xmlblock 'scheme-indent-function 2)
(put 'xmlblockn 'scheme-indent-function 2)
(put 'xmlelt 'scheme-indent-function 'block-indenter)
(put 'scripturl 'scheme-indent-function 2)
(put 'fdscripturl 'scheme-indent-function 2)

(put 'soapenvelope 'scheme-indent-function 2)

(put 'xhtml 'scheme-indent-function 'block-indenter)
(put 'span 'scheme-indent-function 1)
(put 'div 'scheme-indent-function 1)
(put 'table* 'scheme-indent-function 1)
(put 'p 'scheme-indent-function 'block-indenter)
(put 'h1 'scheme-indent-function 'block-indenter)
(put 'h2 'scheme-indent-function 'block-indenter)
(put 'h3 'scheme-indent-function 'block-indenter)
(put 'h4 'scheme-indent-function 'block-indenter)
(put 'anchor 'scheme-indent-function 1)
(put 'anchor* 'scheme-indent-function 2)
(put 'p* 'scheme-indent-function 1)
(put 'h1* 'scheme-indent-function 1)
(put 'h2* 'scheme-indent-function 1)
(put 'tr* 'scheme-indent-function 1)
(put 'th* 'scheme-indent-function 1)
(put 'td* 'scheme-indent-function 1)
(put 'ul 'scheme-indent-function 1)
(put 'li 'scheme-indent-function 1)
(put 'ul* 'scheme-indent-function 1)
(put 'li* 'scheme-indent-function 1)
(put 'vistoggle 'scheme-indent-function 1)
(put 'form 'scheme-indent-function 1)

(put 'pick 'scheme-indent-function 1)
(put 'reject 'scheme-indent-function 1)
(put 'modify-frame 'scheme-indent-function 1)
(put 'frame-update 'scheme-indent-function 1)
(put 'find-frames 'scheme-indent-function 1)
(put 'frame-create 'scheme-indent-function 1)
(put 'index-frame 'scheme-indent-function 2)

(put '%watch 'scheme-indent-function 1)
(put 'deluge%watch 'scheme-indent-function 1)
(put 'detail%watch 'scheme-indent-function 1)
(put 'debug%watch 'scheme-indent-function 1)
(put 'info%watch 'scheme-indent-function 1)
(put 'notice%watch 'scheme-indent-function 1)
(put 'warn%watch 'scheme-indent-function 1)
(put 'always%watch 'scheme-indent-function 1)
(put 'saveoutput 'scheme-indent-function 1)

(put 'extdb/proc 'scheme-indent-function 1)

;;;; Evaluating expressions in modules

(defvar knoc-module)
(defconst in-module-regexp
  "(in-module +'\\(\\(\\w\\|[/$.-_]\\)+\\)")

(defun knoc-get-module-name ()
  "Returns the module name specified in the current buffer"
  (if (and (boundp 'knoc-module) knoc-module) knoc-module
    (save-excursion
      (goto-char (point-min))
      (let* ((pos (search-forward-regexp in-module-regexp (point-max) t))
	     (name
	      (if pos
		  (buffer-substring (match-beginning 1) (match-end 1))
		"")))
;	(if pos
;	    (message "Search found module name %s at %d" name pos)
;	  (message "Search failed to find module name"))
	(if pos (setq knoc-module name))
	name))))
(defun knoc-process () (scheme-proc))

(defun knoc-send-region (start end)
  (let ((module (knoc-get-module-name))
	(process (knoc-process)))
    (if (not (equal module ""))
	(message "Sending %d characters into the %s module"
		 (- end start) module))
    (if (not (equal module ""))
	(comint-send-string
	 process (format "(within-module '%s\n" module)))
    (comint-send-region process start end)
    (if (not (equal module ""))
	(comint-send-string process ")\n")
      (comint-send-string process "\n"))))

(defun knoc-send-definition ()
  "Send the current definition to the inferior Scheme process."
  (interactive)
  (save-excursion
   (end-of-defun)
   (let ((end (point)))
     (beginning-of-defun)
     (knoc-send-region (point) end))))

(defun scheme-send-last-sexp ()
  "Send the previous sexp to the inferior Scheme process."
  (interactive)
  (knoc-send-region (save-excursion (backward-sexp) (point)) (point)))

(defun knoc-sender ()
  (interactive)
  (if mark-active (knoc-send-region (region-beginning) (region-end))
    (knoc-send-definition)))

(defun split-command-line (string)
  (let ((where (string-match "[ \t]" string)))
    (cond ((null where) (list string))
	  ((not (= where 0))
	   (cons (substring string 0 where)
		 (split-command-line (substring string (+ 1 where)
						(length string)))))
	  (t (let ((pos (string-match "[^ \t]" string)))
	       (if (null pos)
		   nil
		 (split-command-line (substring string pos
						(length string)))))))))

(defvar comint-prompt-read-only nil)
(defvar comint-output-read-only nil)

(defun make-output-read-only (text)
  "Add to comint-output-filter-functions to make stdout read only"
  (when comint-output-read-only
    (let ((inhibit-read-only t)
	  (output-end (process-mark (get-buffer-process (current-buffer)))))
      (save-excursion 
	(progn (goto-char output-end)
	       (beginning-of-line)
	       (setq output-end (point))))
      (put-text-property comint-last-output-start output-end 'read-only t))))
(add-hook 'comint-output-filter-functions 'make-output-read-only)

;;; Running an knoc

(defvar knoc-program "knoc")
(defvar knoc-startup nil)
(defvar knoc-cmdline nil)

(defvar knoc-mode-hooks '())

(autoload 'comint-check-proc "comint")

(defun knoc (cmd)
  "Run an inferior KNO scheme process, input and output via buffer *knoc*.
With an arguments, prompts for a command and arguments to use.
If there is a process already running in `*knoc*', switch to that buffer.
Runs the hooks `inferior-scheme-mode-hook' \(after the `comint-mode-hook' is
run). \(Type \\[describe-mode] in the process buffer for a list of commands.)"
  (interactive
   (list (if current-prefix-arg
	     (read-string "Run knoc: "
			  (or knoc-cmdline
			      knoc-program))
	   (or knoc-cmdline
	       knoc-program))))
  (let ((bufname (or (and scheme-buffer
			  (get-buffer-window scheme-buffer)
			  scheme-buffer)
		     "*knoc*"))
	(comint-arg "knoc"))
    (if (equal major-mode (intern "inferior-scheme-mode"))
	(progn (setq bufname (buffer-name (current-buffer)))
	       (setq comint-arg bufname)
	       (if (equal (substring comint-arg 0 1) "*")
		   (setq comint-arg (substring comint-arg 1)))
	       (if (equal (substring comint-arg -1) "*")
		   (setq comint-arg (substring comint-arg 0 -1)))))
    (if (not (comint-check-proc bufname))
	(let ((cmdlist (split-command-line cmd)))
	  (set-buffer (apply 'make-comint comint-arg (car cmdlist)
			     nil (cdr cmdlist)))
	  (inferior-scheme-mode)))
    (setq scheme-program-name cmd)
    (setq scheme-buffer bufname)
    (setq knoc-cmdline cmd)
    (pop-to-buffer bufname)
    (setq comint-prompt-regexp "^#|[^>]+>|")
    (run-hooks 'knoc-mode-hooks)
    ;; (message "Sending '%s'" knoc-startup)
    (when knoc-startup
      (comint-send-string (scheme-proc) (format "%s\n" knoc-startup)))))

(defun fdstartup (string)
  (interactive "sStartup expressions: ")
  (setq-local knoc-startup string))

;;; Defining a mode hook to define knoc-sender

(defun knoc-scheme-mode-hook ()
  (interactive)
  (local-set-key "\e\C-m" 'knoc-sender)
  (setq undo-limit 32)
  (font-lock-add-keywords 'scheme-mode *kno-keywords*))
(add-hook 'scheme-mode-hook 'knoc-scheme-mode-hook)

(provide 'knoc)
;;; knoc.el ends here

